### PR TITLE
Multiply matrix by huge data

### PR DESCRIPTION
**Pull request by resatomo**
Multiply a matrix by long data in each batch to save memory.

### CODE REVIEW

File: aopy/utils/base.py [hunk 1]
```diff
@@ -765,4 +765,39 @@ def convert_port_number(port_number, datatype='ap'):
...
+    Multiply a matrix to data in each batch to save memory. The result is saved in save_path.
+    This function can be used to multiply an inverse matrix by spike band time series.
+    
+    Args:
+        data (nt, nch): neural data. This should be a memory mapping array.
+        mat (anysize, nch): matrix to multiply by data
+        save_path (str): file path to save destriped lfp data
+        sample_rate (float): sampling rate in Hz
+        scale (float, optional): Scaling factor to multiply by data. 1/200 is necessary for whitened data in kilosort4. default is 1.
+        batch_length (int, optional): batch size in second. default is 2 seconds.
```
Comment (rcanfield2) [inline aopy/utils/base.py:781]: Similar to the other pull request, I think there should at least be an option to define the batch size based on available memory.

File: aopy/utils/base.py [hunk 2]
```diff
@@ -765,4 +765,39 @@ def convert_port_number(port_number, datatype='ap'):
...
+    '''
+    batch_size = int(sample_rate * batch_length)
+    n_samples, n_channels = data.shape
+    Nbatches = np.ceil(n_samples/batch_size).astype(int)
+
+    # Multiply matrix in each batch to save memory
+    x = np.memmap(save_path, dtype=dtype, mode='w+', shape=(n_samples, n_channels))
+
+    for ibatch in range(Nbatches):
+        if ibatch < Nbatches-1:
```
Comment (rcanfield2) [inline aopy/utils/base.py:796]: I don't think the if/else statement is required.

File: aopy/utils/base.py [hunk 3]
```diff
@@ -765,4 +765,34 @@ def convert_port_number(port_number, datatype='ap'):
...
+    
+    '''
+        
+    n_samples, n_channels = data.shape
+    batch_size = int (max_memory_gb*1e9 / (n_channels*np.dtype(dtype).itemsize))
+    Nbatches = np.ceil(n_samples/batch_size).astype(int)
+
+    # Multiply matrix in each batch to save memory
+    x = np.memmap(save_path, dtype=dtype, mode='w+', shape=(n_samples, n_channels))
+
```
Comment (rcanfield2) [inline aopy/utils/base.py:794]: can you add the min batch size argument like you did in the other pull request?

Comment (resatomo) [reply aopy/utils/base.py:794]: min batch size is not necessary for this function because this is just multiplication between 2 matrices.

**Review by rcanfield2 — APPROVED** [review]

**Pull request merged**